pkg/proxy/shackles.go

```diff
@@ -0,0 +1,109 @@
+package proxy
+
+import (
+	"net/http"
+	"net/http/httputil"
+
+	stats "github.com/hellofresh/stats-go"
+)
+
+const (
+	statsSectionRoundTrip = "round"
+)
+
+// InLink interface for inbound request plugins
+type InLink func(req *http.Request) (*http.Request, error)
+
+// OutLink interface for outbound request plugins
+type OutLink func(req *http.Request, res *http.Response) (*http.Response, error)
+
+// InChain typed array for inbound plugin sequence
+type InChain []InLink
+
+// OutChain typed array for outbound plugin sequence
+type OutChain []OutLink
+
+// Shackles construct holding plugin sequences
+type Shackles struct {
+	statsClient stats.StatsClient
+	inbound     InChain
+	outbound    OutChain
+}
+
+// NewInChain variadic constructor for inbound plugin sequence
+func NewInChain(in ...InLink) InChain {
+	return append(([]InLink)(nil), in...)
+}
+
+// NewOutChain variadic constructor for outbound plugin sequence
+func NewOutChain(out ...OutLink) OutChain {
+	return append(([]OutLink)(nil), out...)
+}
+
+// RoundTrip provides the Transport.RoundTrip function to handle requests the proxy receives
+func (s *Shackles) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	timing := s.statsClient.BuildTimeTracker().Start()
+
+	// apply inbound request plugins (if any)
+	req, err = s.applyInboundLinks(req)
+	if err != nil {
+		return nil, err
```

Comment:
We must monitor errors like this - currently you're monitoring (using stats) only http-request fail, but any of the errors here make proxy request fail - `s.applyInboundLinks()`,`httputil.DumpResponse()`, `s.applyOutboundLinks()`